### PR TITLE
Bugfixes

### PR DESCRIPTION
**Pull request by skaes**
Hi,

I had started using your gem in one of my projects, but found that many devices are not properly recognized. I have found out that about 70% of all tests of the original user agent fixtures fail. Which is really sad.

The problems are caused by flattening all regular expressions from the regex yaml files and using this flat list. This is wrong; in particular
* the devices yaml files contain a two level regex structure, both of which need to match for a succesful detection
* the device/televisions.yml file must not be used unless the user agent matches a regular expression containing the string "HbbTV/"
* the ruby device detector needs additional ruby logic to match the piwik detector

In order to fix all these problems, I have added the whole set of fixtures and test which are used in the piwik project. I have also rearranged the regex yaml file structure to correspond to the piwik project and added a rake task to simplify updating the regex db.

Please let me know what you think.

### CODE REVIEW

**Comment by yagooar** [comment]
Wow, this PR is huge!

This PR definitely makes sense as adds the piwik fixtures, thus ensures the accuracy of detection.
I'll have to go through it slowly, so please give me some time.

**Comment by skaes** [comment]
Your welcome.

File: Gemfile [hunk 1]
```diff
@@ -2,3 +2,7 @@ source 'https://rubygems.org'
 
 # Specify your gem's dependencies in device_detector.gemspec
 gemspec
+
+if RUBY_VERSION >= "2.0" && defined?(RUBY_ENGINE) && RUBY_ENGINE == "ruby"
+  gem 'byebug'
```
Comment (yagooar) [inline Gemfile]: Having `pry` as a development dependency, do we need `byebug`?

File: lib/device_detector.rb [hunk 2]
```diff
@@ -44,7 +41,55 @@ def device_name
...
+    # which was tablet only, was published too late, there were a
+    # bunch of tablets running with 2.x With 4.0 the two trees were
+    # merged and it is for smartphones and tablets
+    #
+    # So were are expecting that all devices running Android < 2 are
+    # smartphones Devices running Android 3.X are tablets. Device type
+    # of Android 2.X and 4.X+ are unknown
+    if t.nil? && os.short_name == 'AND' && os.full_version && !os.full_version.empty?
+      if os.full_version < "2"
+        t = "smartphone"
```
Comment (yagooar) [inline lib/device_detector.rb]: Please, use single / double quotes consistently.

File: lib/device_detector.rb [hunk 3]
```diff
@@ -44,7 +41,55 @@ def device_name
...
+
+    # Android up to 3.0 was designed for smartphones only. But as 3.0,
+    # which was tablet only, was published too late, there were a
+    # bunch of tablets running with 2.x With 4.0 the two trees were
+    # merged and it is for smartphones and tablets
+    #
+    # So were are expecting that all devices running Android < 2 are
+    # smartphones Devices running Android 3.X are tablets. Device type
+    # of Android 2.X and 4.X+ are unknown
+    if t.nil? && os.short_name == 'AND' && os.full_version && !os.full_version.empty?
```
Comment (yagooar) [inline lib/device_detector.rb]: Wouldn't this be enough:

```ruby
if t.nil? && os.short_name == 'AND' && !os.full_version.empty?
```

File: lib/device_detector.rb [hunk 4]
```diff
@@ -44,7 +41,55 @@ def device_name
...
+    # all Windows 8 touch devices are tablets.
+    if t.nil? && touch_enabled? &&
+       (os.short_name == 'WRT' || (os.short_name == 'WIN' && os.full_version && os.full_version >= '8'))
+      t = "tablet"
+    end
+
+    # set device type to desktop for all devices running a desktop os that were not detected as an other device type
+    if t.nil? && os.desktop?
+      t = "desktop"
+    end
```
Comment (yagooar) [inline lib/device_detector.rb]: I think that the type detecting logic does belong in the `Device#type` method. The file `lib/device_detector.rb` is more of a proxy / public interface of the gem.

File: lib/device_detector/device.rb [hunk 5]
```diff
@@ -10,50 +25,66 @@ def name
...
+          end
+        end
+        regex
+      end
+    end
+
+    def hbbtv?
+      @regex_hbbtv ||= build_regex('HbbTV/([1-9]{1}(\.[0-9]{1}){1,2})')
+      user_agent =~ @regex_hbbtv
+    end
```
Comment (yagooar) [inline lib/device_detector/device.rb]: Why not memoize the whole method an avoid re-matching the regex alltogether?

File: lib/device_detector/device.rb [hunk 6]
```diff
@@ -10,50 +25,66 @@ def name
...
-    def parse_regexes(raw_regexes, device = nil)
-      raw_regexes.map { |base, nest|
+    def matching_regex
+      from_cache([self.class.name, user_agent]) do
+        regex_list = hbbtv? ? regexes_for_hbbtv : regexes_other
+        regex = regex_list.find { |r| user_agent =~ r[:regex] }
+        if regex && regex[:models]
+          model_regex = regex[:models].find { |m| user_agent =~ m[:regex]}
+          if model_regex
+            regex = regex.merge(:regex_model => model_regex[:regex], :model => model_regex[:model])
```
Comment (yagooar) [inline lib/device_detector/device.rb]: Please use the >= 1.9 Hash syntax.

File: lib/device_detector/metadata_extractor.rb [hunk 7]
```diff
@@ -14,23 +14,14 @@ def metadata_string
 
     def extract_metadata
       user_agent.match(regex) do |match_data|
-        replace_metadata_string_with(match_data)
+        metadata_string.gsub(/\$(\d)/) do
+          match_data[$1.to_i].to_s
+        end.strip
```
Comment (yagooar) [inline lib/device_detector/metadata_extractor.rb]: Please, keep this logic in a method, as the method name provides meaningful information about the intention of the `#gsub`.

File: lib/device_detector/os.rb [hunk 8]
```diff
@@ -1,16 +1,152 @@
...
+
+    private
 
-      # This solution won't scale, but for now it will do
-      if raw_version && name == 'Mac'
-        raw_version.split('_').join('.')
+    def os_info
+      from_cache(['os_info', self.class.name, user_agent]) do
+        os_name = NameExtractor.new(user_agent, regex_meta).call
+        if os_name && short = DOWNCASED_OPERATING_SYSTEMS[os_name.downcase]
```
Comment (yagooar) [inline lib/device_detector/os.rb]: Please surround the `short = ...` assignment with parentheses, in order to keep the intention clear.

File: lib/device_detector/os.rb [hunk 9]
```diff
@@ -1,16 +1,152 @@
...
+          os_name = OPERATING_SYSTEMS[short]
+        else
+          short = 'UNK'
+        end
+        { name: os_name, short: short, family: FAMILY_TO_OS[short] }
```
Comment (yagooar) [inline lib/device_detector/os.rb]: A new hash is allocated each time the `#os_info` method is called. Would be great to either memoize it or read it entirely from the cache.

File: lib/device_detector/parser.rb [hunk 10]
```diff
@@ -37,32 +37,43 @@ def filenames
...
     end
 
-    def load_regexes
-      raw_files = filepaths.map { |path| File.read(path) }
+    def load_regexes(file_paths)
+      file_paths.map { |path, full_path| [path, symbolize_keys!(YAML.load_file(full_path))] }
+    end
 
-      raw_files.map { |f| YAML.load(f) }
+    def symbolize_keys!(object)
```
Comment (yagooar) [inline lib/device_detector/parser.rb]: Is the key symbolization really necessary? How does it impact on the performance vs memory allocation in this case?

**Comment by yagooar** [comment]
The file `spec/device_detector/concrete_user_agent_spec.rb` can be deleted, as there are the piwik fixtures.

File: spec/device_detector/bot_fixtures_spec.rb [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    describe File.basename(fixture_file) do
+
+      fixtures = YAML.load_file(fixture_file)
+
+      fixtures.each do |f|
+        user_agent = f["user_agent"]
+        bot = DeviceDetector::Bot.new(user_agent)
+
+        describe user_agent do
+          it "should be a bot" do
```
Comment (yagooar) [inline spec/device_detector/bot_fixtures_spec.rb]: In this project we have been using the "confident" expectation form, e.g. `it 'is a bot'`. Please change it if it's not much asking :)

File: spec/device_detector/client_fixtures_spec.rb [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+      fixtures = YAML.load_file(fixture_file)
+      fixtures.each do |f|
+
+        user_agent = f["user_agent"]
+        client = DeviceDetector::Client.new(user_agent)
+
+        describe user_agent do
+
+          it "should be known" do
+            assert client.known?, "isn't known as a client"
```
Comment (yagooar) [inline spec/device_detector/client_fixtures_spec.rb]: You have been using the `Test::Unit` matchers, which is fine, but the rest of the specs are using `Minitest::Spec`. I already wanted to do the switch, but I think it's better to get this PR reviewed first and open a new one with the change to the test suite.

File: Gemfile (hunk 1 above)
Comment (skaes) [inline Gemfile]: I have never used pry as a debugger. Remove the dependency if you like.

File: lib/device_detector.rb (hunk 3 above)
Comment (skaes) [inline lib/device_detector.rb]: No. os_full_verision can be nil.

File: lib/device_detector.rb (hunk 4 above)
Comment (skaes) [inline lib/device_detector.rb]: The os object os id attached to the device detector, not to the device. You wouldn't want to have a second instance on the device class, would you?

File: lib/device_detector/device.rb (hunk 5 above)
Comment (skaes) [inline lib/device_detector/device.rb]: because memoizing booleans doesn't work well.

File: lib/device_detector/device.rb (hunk 6 above)
Comment (skaes) [inline lib/device_detector/device.rb]: Feel free to change it after merging.

File: lib/device_detector/metadata_extractor.rb (hunk 7 above)
Comment (skaes) [inline lib/device_detector/metadata_extractor.rb]: I don't agree. `replace_meta_data_with` was a horrible name to begin with.

File: lib/device_detector/os.rb (hunk 8 above)
Comment (skaes) [inline lib/device_detector/os.rb]: Makes sense.

File: lib/device_detector/os.rb (hunk 9 above)
Comment (skaes) [inline lib/device_detector/os.rb]: It's already cached. But sure one could do that.

File: lib/device_detector/parser.rb (hunk 10 above)
Comment (skaes) [inline lib/device_detector/parser.rb]: The loaded yaml structures will use a lot less memory as a symbol is a so called immediate value. Access will also be faster as hashing/comparing a symbol is much faster than comparing strings. And I find the code more readable.

File: spec/device_detector/bot_fixtures_spec.rb (hunk 11 above)
Comment (skaes) [inline spec/device_detector/bot_fixtures_spec.rb]: You can change it after merging.

File: spec/device_detector/client_fixtures_spec.rb (hunk 12 above)
Comment (skaes) [inline spec/device_detector/client_fixtures_spec.rb]: Change it if you like.

**Comment by skaes** [comment]
I must say, I'm totally underwhelmed by the feedback.

Your code produced horribly incorrect results, a.k.a. utterly broken.

All of which I fixed with the pull request.

And you criticize single vs. double quotes and ruby hash syntax usage?

**Comment by yagooar** [comment]
> I must say, I'm totally underwhelmed by the feedback.

Well, to begin with, I wasn't finished with the review yet. I first took a look at the surface stuff which doesn't require great tampering with the code.

> Your code produced horribly incorrect results, a.k.a. utterly broken.

There is only finite time we have. We were still iterating on the device detecting stuff, so we were happy to receive your PR.

> All of which I fixed with the pull request.

Well, as I said before, thanks a lot. But being a bit nicer wouldn't hurt either, don't you think?

**Comment by skaes** [comment]
Indeed, everyone's time is limited. So is mine. I have put considerable effort into this patch, including testing it on over 1 million unique user agent strings. My changes fix serious issues with your code. 

If you find a relevant problem with the patch, I'm happy to discuss it, and change things.

However, the code passes all the tests and is even 30% faster than before (on the test suite). Don't you think that discussing minor syntactical details feels outright offending to me? Especially since you can easily change all of this stuff to your liking after merging the code?

Look, I have offered you my help by sending the pull request. If you don't think it's worth merging, that's your decision. But it it would be a shame for your users/customers/clients.

**Comment by yagooar** [comment]
> Don't you think that discussing minor syntactical details feels outright offending to me?

No, didn't think about that. You get easily offended, mister.

While I think that this PR is a very important improvement, your manners and passive-aggressive ways do cost a lot of energy, and discussing would cost a lot of time, too. I am going to merge this PR as is now, and do the necessary changes later on.

Thanks again for the PR.

**Comment by skaes** [comment]
Thx for merging. I appreciate it.

**Pull request merged**